### PR TITLE
Purge SSE2/SSE3

### PR DESCRIPTION
**Pull request by GovanifY**
People are going to hate me for this one.

This removes all the dead code used to support SSE2/SSE3 accelerations.

### CODE REVIEW

**Comment by tellowkrinkle** [comment]
Not sure what happened but `GSVector4.h` has a bunch of random spaces added

**Comment by tellowkrinkle** [comment]
Also check the recompilers for checks to `x86caps.hasStreamingSIMD*`

**Comment by GovanifY** [comment]
> 
> 
> Also check the recompilers for checks to `x86caps.hasStreamingSIMD*`

Did that and looked at non GS source too, I kept the SSE2 checks for the "features detected" log though

**Comment by GovanifY** [comment]
> 
> 
> Not sure what happened but `GSVector4.h` has a bunch of random spaces added

Went over the patch and fixed the spacing issues, there were some on GSVector4i.h too

**Comment by lightningterror** [comment]
Keep solution, spu, core changes in separate commits.
If we want to completely wipe sse2 I think I saw a bunch in  x86emitter and ipu for example.

**Comment by GovanifY** [comment]
> Keep solution, spu, core changes in separate commits.

Ack

> If we want to completely wipe sse2 I think I saw a bunch in x86emitter and ipu for example.

I didn't find any relevant use of x86caps or _M_SSE, you'll have to point me to them

**Comment by tellowkrinkle** [comment]
Also `GSDrawScanlineCodeGenerator.x86.cpp` and `GSDrawScanlineCodeGenerator.cpp` have some `m_cpu.has(util::Cpu::tSSE41)`s including a massive disgusting thing in `ReadTexelSSE`

**Comment by GovanifY** [comment]
> Keep solution, spu, core changes in separate commits.

split everything into different commits

File: pcsx2/SPU2/spu2.cpp [hunk 1]
```diff
@@ -50,20 +50,6 @@ u32 lClocks = 0;
 static bool CheckSSE()
```
Comment (Ziemas) [inline pcsx2/SPU2/spu2.cpp:50]: Why not just delete this function, the only usage is in the one below this.

**Comment by GovanifY** [comment]
> 
> 
> Also `GSDrawScanlineCodeGenerator.x86.cpp` and `GSDrawScanlineCodeGenerator.cpp` have some `m_cpu.has(util::Cpu::tSSE41)`s including a massive disgusting thing in `ReadTexel_SSE`

done

File: pcsx2/SPU2/spu2.cpp (hunk 1 above)
Comment (GovanifY) [reply pcsx2/SPU2/spu2.cpp]: done, thanks for the catch, I thought it was a remnant of the plugin API but it turns out I deleted all traces of it!

**Comment by asturur** [comment]
what will change for minimum cpu supported?

**Comment by GovanifY** [comment]
> 
> 
> what will change for minimum cpu supported?

CPUs will now require at least SSE4 to run PCSX2. This means essentially all CPUs released since 2006.

**Comment by tellowkrinkle** [comment]
2008 (Penryn / 45nm Core 2) for Intel
2011 (FX) for AMD

**Comment by GovanifY** [comment]
> 2008 (Penryn / 45nm Core 2) for Intel

Penryn was launched in 2007, although I indeed forgot about AMD delayed SSE adoption

**Comment by tellowkrinkle** [comment]
> I indeed forgot about AMD delayed SSE adoption
They had SSE4a, just not SSE4.1
Too bad SSE4a is kind of useless

**Comment by asturur** [comment]
oh ok seems a decent cut then. Thanks for the recap

**Comment by RedDevilus** [comment]
If someone has SSE2-only CPU then it won't cut it for 1.6 or later for quite alot of games (i know there are exceptions) . Either upgrade the hardware, use an older build like 1.4 or play on the physical console.

**Comment by GovanifY** [comment]
> 
> 
> Keep solution, spu, core changes in separate commits.
> If we want to completely wipe sse2 I think I saw a bunch in x86emitter and ipu for example.

Double checked just in case:

What you saw was yuv2rgb, which _highest optimisation_ is SSE2, to quote sudonym:

```cpp
// Suikoden Tactics FMV speed results: Reference - ~72fps, SSE2 - ~120fps
// An AVX2 version is only slightly faster than an SSE2 version (+2-3fps)
// (or I'm a poor optimiser),
```

x86emitter, on the other end, is _supposed_ to have SSE2 instructions since it's a JIT emitter and that some code still use SSE2 and doesn't have better optimization, so nothing to do there.

**Comment by Topaz-Reality** [comment]
May I stand against this PR?

Many people use processors that do not have SSE4 Support for their emulation needs, either to re-purpose their old computers or because they cannot afford computers with SSE4. It may not seem like it but some people really need to use old technology.

Not every CPU after 2006 have SSE4 either (Cue Pentium E6800, a Q3'2010 CPU without SSE4 support) and I believe it would kick out a lot of users of PCSX2, which includes me unfortunately.

It seems unnecessary to remove something that a lot of people still use. So please reconsider.

**Comment by JordanTheToaster** [comment]
If you still use a CPU without SSE4 at this point it's likely under the minimum regardless so it wouldn't be supported anyway.

**Comment by Topaz-Reality** [comment]
> If you still use a CPU without SSE4 at this point it's likely under the minimum regardless so it wouldn't be supported anyway.

Quite on the contrary. I can run many PS2 games at full speed. Just because a CPU does not have SSE4 support does not mean it cannot run PCSX2 unless you make it so. In fact, SSE2 remains to be the most performant mode when it comes to PCSX2, regardless of the CPU, and as such 'you cannot run it/you have under the minimum' is a stupid argument that is easily disproven.

As far as I am aware, no other emulator has done this and like I said, not many people have the financial incentive to upgrade, or they repurpose their old PCs with emulators, and this PR will kill off what remains of those groups, which is probably around more than a quarter of the users when you look at it, maybe even more than half. 

Please, **reconsider**.

**Comment by JordanTheToaster** [comment]
> > If you still use a CPU without SSE4 at this point it's likely under the minimum regardless so it wouldn't be supported anyway.
> 
> Quite on the contrary. I can run many PS2 games at full speed. Just because a CPU does not have SSE4 support does not mean it cannot run PCSX2 unless you make it so. In fact, SSE2 remains to be the most performant mode when it comes to PCSX2, regardless of the CPU, and as such 'you cannot run it/you have under the minimum' is a stupid argument that is easily disproven.
> 
> As far as I am aware, no other emulator has done this and like I said, not many people have the financial incentive to upgrade, or they repurpose their old PCs with emulators, and this PR will kill off what remains of those groups, which is probably around more than a quarter of the users when you look at it, maybe even more than half.
> 
> Please, **reconsider**.

As far as my understanding goes SSE2/3 for PCSX2 is the worst performant so that is entirely not true and while I have no numbers either to back up the claim I still stand by what I said either use 1.6 or an earlier dev build.

**Comment by tadanokojin** [comment]
> May I stand against this PR?

You may.

> Many people use processors that do not have SSE4 Support for their emulation needs, either to re-purpose their old computers or because they cannot afford computers with SSE4

Many people are starving too and don't have a PC at all, consider yourself blessed you can run 1.6.0. It is very sad that we don't live in a world where everyone has access to a decent CPU for everything but you can't run to the "well there are people who can't!" argument because there are always people who are worse off.

We need to talk numbers, so just to be clear about the amount of users we're potentially talking about here (steam survey):


I don't expect people to have the latest and greatest technologies but we aren't talking about latest and greatest. There's still SSE4, AVX, AVX2 which we still support and AVX512 which we could support in the future.

> It may not seem like it but some people really need to use old technology.

We're well aware but can't lag on dropping it soley for that reason because again, that's always going to be true. If we went by this logic we'd have to support everything forever (at least in sw terms). It's easy enough to say "ah just support it because people use it". It isn't that simple. There are other considerations to be made than the fact that people exist who use old hardware.

> It seems unnecessary to remove something that a lot of people still use. So please reconsider.

Simple fact is we weighed the pros and cons and the cons outweigh the pros. I don't ask for "unnecessary" changes. If I asked someone to do something it's because I think it's necessary. This is no exception.

> Quite on the contrary. I can run many PS2 games at full speed. 

Again the ability to be able to run the emulator at all is not what is at issue here. I'm sure there are games that can run fullspeed on a decade old CPU. Those games are exceptions that prove the rule. I get that KH is a game that you run (and will be able to run on 1.6.0 but that game runs on a toaster) but KH and 2D fighters for a decade old CPU aren't the typical use case for the emulator. Far from it.

> Just because a CPU does not have SSE4 support does not mean it cannot run PCSX2 unless you make it so.

It's not about "can't" it's about practical. You could run PCSX2 on any number of low powered CPU, doesn't mean it's going to be a good experience. We don't stick "you need 1500 STR" along with a bunch of "this won't be good for most games" language in our reqs because we think that PCSX2 won't physically run on something lower.

> In fact, SSE2 remains to be the most performant mode when it comes to PCSX2

This is demonstrably false and doesn't make any sense.

> and as such 'you cannot run it/you have under the minimum' is a stupid argument that is easily disproven.

I can't tell if it's intentional but you're arguing opposite a point no one is making. No one is arguing that it "can't" be done. It's just impractical for most games and we don't support it.

> As far as I am aware, no other emulator has done this and like I said

Guess we'll be the first? There's plenty of examples of emulators making breaking changes for older systems. I don't see how pointing to this and saying no one has done this specific change yet makes it bad to do. In this case GSdx is pretty exceptional as it needs to rely heavily on SIMD. I don't see many other emulators with a SIMD-accelerated, multi-threaded, and jitted scanline CPU rasterizer either.

> which is probably around more than a quarter of the users when you look at it, maybe even more than half.

[citation needed]

**Comment by asturur** [comment]
I m not involved in the project but i am a developer of other things.
For this kind of projects maintanability and sustainability is very important.

If removing old code makes the code more maintanable you remove it, even if you have to loose some users. The ability for the developers to work on the project better outweight the rest.

I want to run the latest version of pcsx2 on an old pentium 4 is not an argument for anything.

Is open source, fork it, revert this pr on top of any new commit, handle the merge conflicts, compile it, you are done.

**Comment by xddxd** [comment]
https://www.ebay.com/itm/Intel-Core-2-Duo-e8200-e8300-e8400-e8500-e8600-wholesale-LGA-775-CPU-processor/174701270974
thank me later.

**Comment by Topaz-Reality** [comment]
> We need to talk numbers, so just to be clear about the amount of users we're potentially talking about here (steam survey):

Steam Surveys cannot be applied to emulation. Not everyone who emulated will be rockin' a 1080 Ti or a Ryzen 7 or somethin', the great deal with emulation is accessibility. And removing SSE2 will just stand against this.

> Simple fact is we weighed the pros and cons and the cons outweigh the pros. I don't ask for "unnecessary" changes. If I asked someone to do something it's because I think it's necessary. This is no exception.

May I learn of these pros and cons so that my mind can be changed?

> Again the ability to be able to run the emulator at all is not what is at issue here. I'm sure there are games that can run fullspeed on a decade old CPU. Those games are exceptions that prove the rule. I get that KH is a game that you run (and will be able to run on 1.6.0 but that game runs on a toaster) but KH and 2D fighters for a decade old CPU aren't the typical use case for the emulator. Far from it.

KH and Soulcalibur are NOT the only games I run, even though it may seem like it. I also run the Crash series, the Sims series, the Sly Cooper series, the God of War series and more. Do not make assumptions based on my modding career please. 
Again, the greatest use case for emulators are proven to be for repurposing old hardware, heck, this even goes for old consoles.

Again, seeing the pros and cons form you guys, my mind may be changed. But until then, I will always thrive to protect backwards compatibility and legacy support.

> This is demonstrably false and doesn't make any sense.

I have ran tests myself on 2 laptops circa 2014. I can post benchmarks whenever I can get a non-impactful video recording solution set.

> I can't tell if it's intentional but you're arguing opposite a point no one is making. No one is arguing that it "can't" be done. It's just impractical for most games and we don't support it.

Quoting one guy who said 'If you still use a CPU without SSE4 at this point it's likely under the minimum regardless so it wouldn't be supported anyway.' above.

> There's plenty of examples of emulators making breaking changes for older systems. I don't see how pointing to this and saying no one has done this specific change yet makes it bad to do.

I can understand cutting support on deprecated APIs (DX9) or unsupported OSes (Win7, Win8) even though I do not support those decisions. But it seems baffling to e that you would cut out a supported architecture (All modern OSes will still run on SSE2) and in fact one that many of your peers still use.

Again, if I can see the pros and cons, my mind may be changed.

> I want to run the latest version of pcsx2 on an old Pentium 4 is not an argument for anything.

Pentium 4? What post are you reading? Many modern Pentiums did not support SSE4 either.

> thank me later.

Thanks for being a certified dickhead. Sorry for breaking professionalism but this had to be said.

**Comment by xddxd** [comment]
You’re welcome. Old hardware should not block progress from being made.

**Comment by tellowkrinkle** [comment]
There's 2 main issues
1. The plan for the [GSdx merge](https://github.com/PCSX2/pcsx2/discussions/4100) involves disabling multiply-built GSdx as we carry out the merge.  GSdx benefits most from SSE4 (AVX/AVX2 affect it much less), so the officially built PCSX2 will definitely drop SSE3 for that period whether we keep it as a build-time option or not
2. We keep accidentally breaking SSE3.  SSE4 has a lot of useful instructions (min/max of integers, pinsr/pextr, movsx/movzx) that we end up adding to new code and not noticing until someone comes in with a sigill later.  Most recently, SSE3 was broken for two months from ee07f860fc1dd7e8ada6c13903bb3a592b8d4864 to f9d96f55a538ce373f1662a5554d8052173940b1 

I'll probably try to maintain SSE2 on my fork through the next official release as I'd like to have a fallback for Mac users (on Windows we can say "use 1.6.0" but there is no 1.6.0 for Mac), and since that'll include the plugin merge, going from there will probably not have too many merge conflicts to deal with for anyone else who wants to have SSE2 support, and anyone who is trying to keep SSE2 alive probably actually has an SSE2 computer they can test on to make sure new changes actually work

**Comment by asturur** [comment]
Again if the project is hard to mantain, you loose the ability to run newer versions entirely, because developers may stop doing the emulator at all. Have developers do what they have to do.

**Comment by Topaz-Reality** [comment]
> Again if the project is hard to mantain, you loose the ability to run newer versions entirely, because developers may stop doing the emulator at all. Have developers do what they have to do.

I am talking with the devs as you are seeing here. Let the conversation run, please.

> We keep accidentally breaking SSE3. SSE4 has a lot of useful instructions (min/max of integers, pinsr/pextr, movsx/movzx) that we end up adding to new code and not noticing until someone comes in with a sigill later. Most recently, SSE3 was broken for two months.

I see... Well, that is understandable. But isn't it possible to shield all the SSE4 stuff from SSE2/SSE3 stuffs? Excuse me if I am talking nonsense but I am just asking because you can do this very thing to shield x86 code from x86_64 code.

> I'll probably try to maintain SSE2 on my fork through the next official release as I'd like to have a fallback for Mac users

My fork will also keep SSE2 support. But all of the backwards and legacy support aside, I am currently trying to find a way to continue #4199 because I, again, do not have a SSE4-compatible system on hand. If it was not for the GSdx merge this would have been simple, just edit the plugin itself, but it will be much more difficult after the merge.

Perhaps I can find a workaround to this but we will see.

**Comment by fantesykikachu** [comment]
> I see... Well, that is understandable. But isn't it possible to shield all the SSE4 stuff from SSE2/SSE3 stuffs? Excuse me if I am talking nonsense but I am just asking because you can do this very thing to shield x86 code from x86_64 code.

By shield you mean duplicate, and yes in theory they could keep it but it would be horrible to maintain. It is the same way with x86 and x86_64, right now there is a lot of duplication but as soon as x86_64 is fully working the plan is to drop x86 (32bit) support, just like many other emulators did like five years ago.

**Comment by tadanokojin** [comment]
> Steam Surveys cannot be applied to emulation

Then nothing can be used because there are no stats that applies specifically to ps2 emulation (at least without telepathy, which is non-starter afaic) and we effectively have no data to rely on. Did you see the qualifier "potentially" in that sentence? I don't mean to be condescending but I'm not sure you realized that word is doing the work of telling you that this isn't meant to be interpreted as 1:1 with emulation users.

It's a decent sampling of gamers more broadly but even so I'm accounting for the fact that the sample might be slightly bias and I still don't think SSE2 is worth it. I think it's fair to recognize the overlap between this two groups of computer users and account for any potential bias in the data rather than just dismiss it outright.

Additionally, you're welcome to present other stats you feel are more accurately represent out user-base.

> Not everyone who emulated will be rockin' a 1080 Ti or a Ryzen 7 or somethin', 

I didn't suggest everyone does and neither does steam. AMD is barely a quarter of the market let alone a Ryzen 7 and a 1080ti isn't even in the top 10. I'm assuming you're exaggerating for effect but I'm not really sure what you're trying to suggest here.

> the great deal with emulation is accessibility. And removing SSE2 will just stand against this. 

You need to draw the line somewhere and It's clearly balancing act. Bring more than an absolutest stance about accessibility. Where is the line and why?

> KH and Soulcalibur are NOT the only games I run

Nothing about that suggested they were the only games you run, just that they were games I'm aware you run which I preemptively addressed before they came up. KH in particular I wanted to address preemptively because it's an exceptional case that I suspected would come up given the nature of the conversation.

I wasn't even aware that you played SC specifically, I said "2D fighting games" because that seemed more charitable than the other 2D games I could think of as an example (visual novels). SC is a 3D fighting game last I checked.

> Quoting one guy who said 'If you still use a CPU without SSE4 at this point it's likely under the minimum regardless so it wouldn't be supported anyway.' above.

That's not saying you can't use it. It's saying we likely won't provide support for it because it's too under-powered, which is true. I'm going to be charitable and say you probably just didn't understand what they were saying.

> I can understand cutting support on deprecated APIs (DX9) or unsupported OSes (Win7, Win8) even though I do not support those decisions. But it seems baffling to e that you would cut out a supported architecture (All modern OSes will still run on SSE2) and in fact one that many of your peers still use.

We're not developing an OS or anything close to it. This is a confused point, d3d9 isn't deprecated afaik and I'm pretty sure there are operating systems that require at least SSE4 (MacOS and some linux distros I'm pretty sure).

**Comment by seta-san** [comment]
I feel for you. Bit we're talking about computers 10 to 15 years old. We dump computers by the dozens half as old.. I think you said you had a 775 socket computer. Don't core 2 quads supports sse4?

**Comment by tellowkrinkle** [comment]
> My fork will also keep SSE2 support. But all of the backwards and legacy support aside, I am currently trying to find a way to continue #4199 because I, again, do not have a SSE4-compatible system on hand. If it was not for the GSdx merge this would have been simple, just edit the plugin itself, but it will be much more difficult after the merge.

I usually have two branches for PRs, one with MacOS patches and one without

I work on the one with patches, and when I'm done I `git cherry-pick` the commits onto the one I actually PR

**Comment by asturur** [comment]
> > Again if the project is hard to mantain, you loose the ability to run newer versions entirely, because developers may stop doing the emulator at all. Have developers do what they have to do.
> 
> I am talking with the devs as you are seeing here. Let the conversation run, please.
> 
I do not think this is the way to facilitate any running conversation.  You should use any private channel with developers if you can't stand other points of view.

**Review by lightningterror — APPROVED** [review]
I think it looks good, could use some testing just to make sure nothing broke.
Maybe Tellow or Li can double check if something got left out.

**Comment by GovanifY** [comment]
@Mrlinkwii tested the PR and it seemed fine, @tellowkrinkle care to double check the PR in case?

File: plugins/GSdx/GSVector4.h [hunk 2]
```diff
@@ -342,54 +316,22 @@ class alignas(16) GSVector4
...
-		#if 0//_M_SSE >= 0x501
-
-		return GSVector4(_mm_fnmsub_ps(m, a, b));
-		
-		#else
-
 		return -b - *this * a;
-
-		#endif
 	}
```
Comment (tadanokojin) [inline plugins/GSdx/GSVector4.h]: do we need to ditch the nopped out FMA instructions?

Comment (tadanokojin) [reply plugins/GSdx/GSVector4.h]: nvm I see you killed it in a later commit.

**Comment by RedDevilus** [comment]
https://github.com/PCSX2/pcsx2/blob/master/pcsx2/Docs/PCSX2_FAQ.md#question-5-what-are-the-pc-requirements-to-use-pcsx2
Minimum: SSE2 will turn into SSE4 as well.

**Comment by tellowkrinkle** [comment]
Looks good to me, though I'd prefer if you reordered the commits so that they all build (build system changes first)

**Review by tadanokojin — COMMENTED** [review]
GS stuff looks okay to me, I think you should merge the two GS commits.

**Comment by GovanifY** [comment]
> 
> 
> Looks good to me, though I'd prefer if you reordered the commits so that they all build (build system changes first)
> 
> Edit: Also rebase the part where you un-delete things you deleted together so it remains unchanged by any of the commits



> 
> 
> Looks good to me, though I'd prefer if you reordered the commits so that they all build (build system changes first)
> 
> Edit: Also rebase the part where you un-delete things you deleted together so it remains unchanged by any of the commits

done



> 
> 
> GS stuff looks okay to me, I think you should merge the two GS commits.

done

**Review by RedDevilus — APPROVED** [review]
LGTM.

**Pull request merged**